### PR TITLE
feat(backend/copilot): add include_graph option to find_library_agent

### PR DESCRIPTION
**Pull request by majdyz**
## Why

The copilot's `edit_agent` tool requires the LLM to provide a complete agent JSON (all nodes + links), but the LLM had **no way to see the current graph structure** before editing. It was editing blindly — guessing/hallucinating the entire node+link structure and replacing the graph wholesale.

## What

- Add `include_graph` boolean parameter (default `false`) to the existing `find_library_agent` tool
- When `true`, each returned `AgentInfo` includes a `graph` field with the full graph JSON (nodes, links, `input_default` values)
- Update the agent generation guide to instruct the LLM to always fetch the current graph before editing

## How

- Added `graph: dict[str, Any] | None` field to `AgentInfo` model
- Added `_enrich_agents_with_graph()` helper in `agent_search.py` that calls the existing `get_agent_as_json()` utility to fetch full graph data
- Threaded `include_graph` parameter through `find_library_agent` → `search_agents` → `_search_library`
- Updated `agent_generation_guide.md` to add an "if editing" step that fetches the graph first

No new tools introduced — reuses existing `find_library_agent` with one optional flag.

## Test plan

- [x] Unit tests: 2 new tests added (`test_include_graph_fetches_nodes_and_links`, `test_include_graph_false_does_not_fetch`)
- [x] All 7 `agent_search_test.py` tests pass
- [x] All pre-commit hooks pass (lint, format, typecheck)
- [ ] Verify copilot correctly uses `include_graph=true` before editing an agent (manual test)

### CODE REVIEW

File: autogpt_platform/backend/backend/copilot/tools/agent_search.py [hunk 1]
```diff
@@ -196,6 +204,18 @@ async def _search_library(
     )
 
 
+async def _enrich_agents_with_graph(agents: list[AgentInfo], user_id: str) -> None:
```
Comment (majdyz) [inline autogpt_platform/backend/backend/copilot/tools/agent_search.py]: 🤖 🟠 **Should Fix**: Sequential graph fetching — this loops through agents one-by-one with `await`. If the user calls `find_library_agent(include_graph=true)` with no query (list all), this could try to fetch graphs for up to 50 agents sequentially, causing significant latency.

Suggest using `asyncio.gather` to fetch in parallel, or capping the number of graphs fetched (e.g., max 5-10).

File: autogpt_platform/backend/backend/copilot/tools/agent_search.py [hunk 2]
```diff
@@ -196,6 +204,18 @@ async def _search_library(
...
+    """Fetch and attach full graph JSON (nodes + links) to each agent in-place."""
+    for agent in agents:
+        graph_id = agent.graph_id
+        if not graph_id:
+            continue
```
Comment (majdyz) [inline autogpt_platform/backend/backend/copilot/tools/agent_search.py]: 🤖 🟡 **Nice to Have**: Agreeing with Sentry's comment — `get_agent_as_json` can return `None` (agent not found), which silently assigns `None` to `agent.graph`. This is indistinguishable from the default state. Consider logging a warning when it returns `None` so graph-not-found issues are observable:

```python
graph_json = await get_agent_as_json(graph_id, user_id)
if graph_json is None:
    logger.warning(f"Graph not found for agent {graph_id}")
agent.graph = graph_json
```

File: autogpt_platform/backend/backend/copilot/tools/agent_search_test.py [hunk 3]
```diff
@@ -168,3 +175,75 @@ async def test_uuid_lookup_found_by_graph_id(self):
...
+                query=agent_id,
+                source="library",
+                session_id="test-session",
+                user_id=_TEST_USER_ID,
+                include_graph=False,
+            )
+
+        assert isinstance(response, AgentsFoundResponse)
+        assert response.agents[0].graph is None
+        mock_get_json.assert_not_awaited()
```
Comment (majdyz) [inline autogpt_platform/backend/backend/copilot/tools/agent_search_test.py]: 🤖 🟡 **Nice to Have**: Consider adding a test for the failure path — when `get_agent_as_json` raises an exception or returns `None`. This would verify that the error is handled gracefully and doesn't break the response.

File: autogpt_platform/backend/backend/copilot/tools/agent_search.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_search.py]: Fixed in 49432af — now logs a warning when `get_agent_as_json` returns `None`, and added a test for this case.

File: autogpt_platform/backend/backend/copilot/tools/agent_search.py (hunk 1 above)
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_search.py]: Fixed in 49432af — now uses `asyncio.gather` for parallel fetching and caps at `_MAX_GRAPH_FETCHES = 10`.

File: autogpt_platform/backend/backend/copilot/tools/agent_search.py (hunk 2 above)
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_search.py]: Fixed in 49432af — added explicit `None` check with warning log.

File: autogpt_platform/backend/backend/copilot/tools/agent_search_test.py (hunk 3 above)
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_search_test.py]: Fixed in 49432af — added `test_include_graph_handles_fetch_failure` (exception path) and `test_include_graph_handles_none_return` (None return path). 9 tests now pass.

File: autogpt_platform/backend/backend/copilot/tools/agent_search.py [hunk 4]
```diff
@@ -196,6 +205,27 @@ async def _search_library(
...
+
+    async def _fetch(agent: AgentInfo) -> None:
+        try:
+            graph_json = await get_agent_as_json(agent.graph_id, user_id)  # type: ignore[arg-type]
+            if graph_json is None:
+                logger.warning(f"Graph not found for agent {agent.graph_id}")
+            agent.graph = graph_json
+        except Exception as e:
+            logger.warning(f"Failed to fetch graph for agent {agent.graph_id}: {e}")
+
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_search.py:257]: Fixed in 9cf8404: replaced `# type: ignore[arg-type]` with an early return guard clause that narrows `agent.graph_id` from `str | None` to `str`.

[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_search.py:257]: The `Graph` Pydantic model (`backend.data.graph.Graph`) has computed properties (`input_schema`, `output_schema`, `has_external_trigger`, etc.) that trigger block registry lookups at serialization time and add redundant data already present in the top-level `AgentInfo` fields.

More importantly, the `edit_agent` tool consumes `agent_json: dict[str, Any]` — the LLM needs a plain dict it can modify and pass back. `get_agent_as_json()` → `graph_to_json()` produces exactly that round-trip format (nodes + links as flat dicts).

Using the typed `Graph` model would couple the response to internal Pydantic model structure, make the response heavier, and the LLM would still need to convert back to dict for `edit_agent`. The existing pattern (`AgentPreviewResponse.agent_json`, `FixResultResponse.fixed_agent_json`) also uses `dict[str, Any]` for the same reason.

**Review by majdyz — COMMENTED** [review]
**Review: LGTM with one fix applied.**

**Changes reviewed:**
- New `include_graph` parameter on `find_library_agent` tool and `search_agents`
- `_enrich_agents_with_graph()` fetches graph JSON via `get_agent_as_json()` with `asyncio.gather`
- `AgentInfo.graph` field added to model + OpenAPI schema
- Agent generation guide updated with `include_graph=true` workflow step
- 5 new test cases covering success, failure, None return, and opt-out paths

**Fix applied (9cf8404):** Replaced `# type: ignore[arg-type]` with a proper guard clause that narrows `agent.graph_id` from `str | None` to `str` before passing to `get_agent_as_json()`.

**No bugs, security issues, or breaking changes found.** The `_MAX_GRAPH_FETCHES = 10` cap and graceful error handling are solid. The `asyncio.gather` pattern is appropriate here since graph fetches are independent I/O-bound operations.

File: autogpt_platform/backend/backend/copilot/tools/agent_search.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_search.py:257]: Updated in 22d05bc — now uses `Graph` type directly instead of `dict[str, Any]`. The field is now `graph: Graph | None` on `AgentInfo`, and the enrichment calls `graph_db().get_graph()` directly instead of going through `get_agent_as_json()`.

File: autogpt_platform/backend/backend/copilot/tools/agent_search.py [hunk 5]
```diff
@@ -196,6 +205,30 @@ async def _search_library(
...
 
+_MAX_GRAPH_FETCHES = 10
+
+
+async def _enrich_agents_with_graph(agents: list[AgentInfo], user_id: str) -> None:
+    """Fetch and attach full graph JSON (nodes + links) to each agent in-place."""
+    fetchable = [a for a in agents if a.graph_id][:_MAX_GRAPH_FETCHES]
+    if not fetchable:
+        return
+
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_search.py:238]: Fixed in 22d05bc: Added truncation logging to `_enrich_agents_with_graph`. When more agents have `graph_id` than `_MAX_GRAPH_FETCHES` allows, a warning is now logged with the count of skipped agents.

[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_search.py:238]: Already addressed in the latest commit — the function now logs a warning when truncation occurs:

```python
skipped = len(with_graph_id) - len(fetchable)
if skipped > 0:
    logger.warning(
        "include_graph: fetched graphs for %d/%d agents "
        "(_MAX_GRAPH_FETCHES=%d, %d skipped)",
        len(fetchable), len(with_graph_id), _MAX_GRAPH_FETCHES, skipped,
    )
```

The truncation is intentional — fetching 50 graphs in parallel would be expensive. The LLM typically uses this with a specific agent_id (1 result), not a broad search.

File: autogpt_platform/backend/backend/copilot/tools/agent_search.py [hunk 6]
```diff
@@ -196,6 +205,49 @@ async def _search_library(
...
+    """
+    with_graph_id = [a for a in agents if a.graph_id]
+    fetchable = with_graph_id[:_MAX_GRAPH_FETCHES]
+    if not fetchable:
+        return
+
+    async def _fetch(agent: AgentInfo) -> None:
+        graph_id = agent.graph_id
+        if not graph_id:
+            return
```
Comment (majdyz) [inline autogpt_platform/backend/backend/copilot/tools/agent_search.py:244]: 🤖 🔵 **Nit**: This guard is redundant — `fetchable` is already filtered to agents with `graph_id` (line 218). Harmless but can be removed for clarity.

Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_search.py:244]: Fixed in 30f6989 — removed the redundant guard.

File: autogpt_platform/backend/backend/copilot/tools/agent_search.py [hunk 7]
```diff
@@ -196,6 +205,46 @@ async def _search_library(
...
+    async def _fetch(agent: AgentInfo) -> None:
+        try:
+            graph_json = await get_agent_as_json(
+                agent.graph_id, user_id  # type: ignore[arg-type]
+            )
+            if graph_json is None:
+                logger.warning(f"Graph not found for agent {agent.graph_id}")
+            agent.graph = graph_json
+        except Exception as e:
+            logger.warning(f"Failed to fetch graph for agent {agent.graph_id}: {e}")
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_search.py]: Fixed in 4c71eea — extracted `graph_id` with an early-return guard so the type checker narrows correctly, removing the `# type: ignore[arg-type]` suppressor.

File: autogpt_platform/backend/backend/copilot/tools/agent_search.py [hunk 8]
```diff
@@ -196,6 +205,49 @@ async def _search_library(
...
+    async def _fetch(agent: AgentInfo) -> None:
+        graph_id = agent.graph_id
+        if not graph_id:
+            return
+        try:
+            graph = await get_graph_db().get_graph(
+                graph_id, version=None, user_id=user_id
+            )
+            if graph is None:
+                logger.warning(f"Graph not found for agent {graph_id}")
```
Comment (majdyz) [inline autogpt_platform/backend/backend/copilot/tools/agent_search.py]: 🤖 🟡 Nice to Have: The f-strings in `logger.warning()` calls (lines 232, 235) bypass lazy formatting. Consider using `logger.warning("Graph not found for agent %s", graph_id)` style. Not performance-critical since these are on exceptional warning paths, but it's the standard convention for logging.

Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_search.py]: Fixed in 98ae9df: replaced f-strings with %-style lazy formatting in logger.warning calls.

**Comment by majdyz** [comment]
**Review Summary (3 rounds)**

**Round 1-3:** Clean. No bugs, security issues, or breaking changes found.

**Status:**
- 0 unresolved external review threads (1 resolved during this review — style nit already fixed)
- CI: all checks passing
- Code quality: parallel graph fetching with asyncio.gather, proper error handling, cap on fetches (_MAX_GRAPH_FETCHES=10), comprehensive test coverage (4 new tests)

**Comment by majdyz** [comment]
All blockers and should-fix items addressed in 43477ed:

**Blocker fix (CRITICAL security):**
- `get_graph()` now called with `for_export=True` — `stripped_for_export()` filters `credentials`, `api_key`, `password`, `token`, `secret` fields from `input_default` before graph data reaches the LLM context

**Should-fix items:**
- `version=agent.graph_version` instead of `version=None` — fetches the active version only, not draft/unpublished versions
- `asyncio.timeout(15)` wraps `asyncio.gather` — prevents indefinite blocking on hung DB connections
- `graph_db()` resolved once before the gather instead of per-coroutine
- Dropped `get_graph_db` alias in favor of `graph_db` to match codebase convention

All 9 tests pass. Re-requesting review.

/review

File: autogpt_platform/backend/backend/copilot/tools/agent_search.py [hunk 9]
```diff
@@ -196,6 +204,62 @@ async def _search_library(
...
+                user_id=user_id,
+                for_export=True,
+            )
+            if graph is None:
+                logger.warning("Graph not found for agent %s", graph_id)
+            agent.graph = graph
+        except Exception as e:
+            logger.warning("Failed to fetch graph for agent %s: %s", graph_id, e)
+
+    async with asyncio.timeout(_GRAPH_FETCH_TIMEOUT):
```
Comment (majdyz) [inline autogpt_platform/backend/backend/copilot/tools/agent_search.py]: 🤖 🟠 **Should Fix**: Unhandled `TimeoutError` from `asyncio.timeout()` will crash the tool call.

If graph fetches take longer than 15 seconds, `asyncio.timeout()` raises `TimeoutError` which propagates up through `_search_library` and crashes the entire tool response. Other tools in this codebase (sandbox, agent_browser) explicitly catch `asyncio.TimeoutError`.

The timeout should be caught and the function should return gracefully (agents with partial/no graph data), since graph enrichment is optional and shouldn't prevent the agent list from being returned.

Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_search.py]: 🤖 Fixed in 862906fdd: Wrapped `asyncio.timeout` in try/except TimeoutError so graph enrichment degrades gracefully. Agents are still returned with whatever graph data was fetched before the timeout.

**Comment by majdyz** [comment]
## Test Report (Round 2): `include_graph` option for `find_library_agent`

**Tested on local server** (backend :8006, frontend :3000, user test@test.com)

### Results Summary

| Test | Result |
|------|--------|
| OpenAPI spec has `graph` field on `AgentInfo` | :white_check_mark: PASS |
| Tool schema exposes `include_graph` parameter | :white_check_mark: PASS |
| Parameter flows through `_execute -> search_agents -> _search_library -> _enrich_agents_with_graph` | :white_check_mark: PASS |
| Security: `for_export=True` strips secrets | :white_check_mark: PASS |
| Safety: max 10 fetches, 15s timeout, TimeoutError handling | :white_check_mark: PASS |
| Unit tests: all 9 agent_search tests pass | :white_check_mark: PASS |
| Additional tests (helpers: 23/23, tool_adapter: 41/41) | :white_check_mark: PASS |

### Details

**Code review:** The `include_graph` parameter correctly flows through the full call chain:
`FindLibraryAgentTool._execute()` -> `search_agents()` -> `_search_library()` -> `_enrich_agents_with_graph()`

Key security/safety features:
- `for_export=True` on `get_graph()` strips credentials and secrets before data reaches the LLM context
- `_MAX_GRAPH_FETCHES = 10` caps concurrent graph fetches
- `_GRAPH_FETCH_TIMEOUT = 15s` prevents hanging
- `TimeoutError` and general exceptions caught gracefully (agents still returned without graph)

**API testing:** Created a test agent, confirmed it appears in library. Sent 3 chat sessions requesting `include_graph=true`. The tool was called successfully but the LLM did not pass `include_graph=true` (hallucinated the param doesn't exist). This is an LLM behavior issue, not a code bug -- the parameter IS correctly registered in the tool schema and SDK.

**Unit tests:** All 5 `include_graph`-specific tests pass:
- `test_include_graph_fetches_graph` -- verifies graph attachment with `for_export=True`
- `test_include_graph_false_skips_fetch` -- verifies skip on default False
- `test_include_graph_handles_fetch_failure` -- graceful degradation
- `test_include_graph_handles_none_return` -- null safety
- Plus TimeoutError handling in `_enrich_agents_with_graph`

**Note:** The CoPilot LLM consistently refused to pass `include_graph=true` in live testing, claiming the parameter doesn't exist. This may warrant adding `include_graph` to the system prompt's tool documentation so the LLM knows about it. Not a blocker for the code changes.

File: autogpt_platform/backend/backend/copilot/tools/agent_search.py [hunk 10]
```diff
@@ -196,6 +204,67 @@ async def _search_library(
...
+            agent.graph = graph
+        except Exception as e:
+            logger.warning("Failed to fetch graph for agent %s: %s", graph_id, e)
+
+    try:
+        async with asyncio.timeout(_GRAPH_FETCH_TIMEOUT):
+            await asyncio.gather(*[_fetch(a) for a in fetchable])
+    except TimeoutError:
+        logger.warning(
+            "include_graph: timed out after %ds fetching graphs", _GRAPH_FETCH_TIMEOUT
```
[the remark this answers is not in the document]
Comment (majdyz) [reply autogpt_platform/backend/backend/copilot/tools/agent_search.py]: Fixed in c9bf3aa: On timeout, all partially-enriched `.graph` fields are now cleared (`a.graph = None`) so callers see consistent all-or-nothing state instead of a mix.

**Comment by majdyz** [comment]
## E2E Test Results - Round 3 (2026-04-02)

**Environment:** localhost:3000/8006, combined-preview-test, worktree AutoGPT2

### 1. Unit Tests: PASS (13/13)
```
poetry run pytest backend/copilot/tools/agent_search_test.py -x -q
.............  13 passed in 0.08s
```

### 2. OpenAPI Spec: PASS
`AgentInfo.graph` field correctly defined as `BaseGraph-Output | null`. Frontend `openapi.json` synced.

### 3. REST API: PASS
`GET /api/library/agents` does **not** expose `include_graph` as a query param (correct -- copilot-only tool parameter). Default responses omit `graph` field.

### 4. CoPilot SSE with `include_graph=true`: PASS
- User prompt: _"List all my agents and include the graph structure"_
- Copilot called `find_library_agent({"include_graph": true})`
- 20 agents returned: **10 with full graph data**, 10 with `graph: null` (truncated at `_MAX_GRAPH_FETCHES=10`)
- Truncation notice surfaced: _"Graph data included for 10 of 20 eligible agents (limit: 10)"_
- Graph objects contain: `id`, `version`, `nodes`, `links`, `input_schema`, `output_schema`, etc.
- **Secret stripping verified** via `for_export=True`: no API keys/passwords in `input_default`

### 5. CoPilot SSE default (no graph): PASS
- User prompt: _"What agents do I have in my library?"_
- Copilot called `find_library_agent({})` -- no `include_graph`
- All 21 agents returned with `"graph": null` -- zero graph fetch overhead

### 6. Agent Generation Guide: PASS
Editing workflow now starts with "fetch graph via `find_library_agent(query=<id>, include_graph=true)`" before making changes. Steps renumbered correctly.

### Summary

| Test | Status |
|------|--------|
| Unit tests (13/13) | :white_check_mark: |
| OpenAPI spec | :white_check_mark: |
| REST API (no include_graph leakage) | :white_check_mark: |
| CoPilot SSE with include_graph=true | :white_check_mark: |
| CoPilot SSE default (no graph) | :white_check_mark: |
| Truncation at MAX_GRAPH_FETCHES=10 | :white_check_mark: |
| Secret stripping (for_export=True) | :white_check_mark: |
| Agent generation guide | :white_check_mark: |

**All tests passed. PR is ready to merge.**

**Review by ntindle — APPROVED** [review]
